### PR TITLE
Sphinx docs

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

Prepare documentation folder to make it build-able with Sphinx (see #960 for reasons why).

I tried to modify one of the original sphinx themes to use a dark background. The results of how it currently looks like are shown in the screenshot below:
Screen Shot 2021-09-20 at 1 54 47 AM

In order to build the docs locally, you'd need to first run `pip install -r requirements.txt`, and then `make html`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [ ] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

WIP for #960

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

**Review by spydon — APPROVED** [review]

File: doc/conf.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# Configuration file for the Sphinx documentation builder.
```
Comment (luanpotter) [inline doc/conf.py:1]: would it be possible to move any build related files into a specific folder? to separate them from the actual markdown

**Comment by luanpotter** [comment]
Great idea! We do need to find a way to wire it up for the website deploy

File: doc/conf.py (hunk 1 above)
Comment (st-pasha) [reply doc/conf.py]: I've put all the config files into the `doc/_conf/` folder.
It works fine locally: you just need to be in that folder when running `make html` command.

File: doc/README.md [hunk 2]
```diff
@@ -53,3 +53,48 @@ You can also check out the
 [awesome flame repository](https://github.com/flame-engine/awesome-flame#articles--tutorials),
 It contains quite a lot of good tutorials and articles written by the community to get you started
 with Flame.
+
+```{toctree}
```
Comment (luanpotter) [inline doc/README.md:57]: This breaks it on github view: 
image

I wonder if we can find a way to make it work both ways. or at least omit it on github

File: .readthedocs.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# [Read the Docs] configuration file
```
Comment (luanpotter) [inline .readthedocs.yaml:1]: can this also be inside the _config folder? or does that break anything?

**Review by luanpotter — COMMENTED** [review]
I think we can merge this and followup on make sure the deploy works later. I just have a couple final comments I left on the PR, plus:

* should we use flame icon & logo instead of Blue Fire? since this is flame-related doc?
* can the static folder be inside the config folder too? (not sure if possible)

File: .readthedocs.yaml (hunk 3 above)
Comment (st-pasha) [reply .readthedocs.yaml:1]: This file can only be in the repository root, or ReadTheDocs won't be able to find it. The only other option is to not have this file at all, and instead set up all configuration options through their web interface. But that's considered more fragile, so [not recommended](https://docs.readthedocs.io/en/stable/config-file/index.html).

File: doc/README.md (hunk 2 above)
Comment (st-pasha) [reply doc/README.md]: I've put the table of contents into a separate file. This way the readme can stay exactly the way it is now, but there is also a new `index.md` file, and new `toc.rst` file.

**Comment by st-pasha** [comment]
> should we use flame icon & logo instead of Blue Fire? since this is flame-related doc?

I thought the change to Blue Fire / blue color was across the board. But if you're saying Flame remains orange/red as it was before, then it's my misunderstanding. I've changed the logo and favicon to use the correct colors now.

**Comment by st-pasha** [comment]
> can the static folder be inside the config folder too? (not sure if possible)

The folder can be renamed/moved, no problem. However, the purpose of this folder is not to provide build configuration, but rather it's a store for static content that will end up being copied into the resulting HTML website. This typically includes things like stylesheets, images, scripts, etc. It is pretty much equivalent to the `assets` folder in Android/iOS project. Perhaps we should rename it as such?

It would probably make sense to move the current `images` folder inside that folder -- but that's better to do as a separate PR.

**Comment by st-pasha** [comment]
After all the tweaks, the doc now renders like this:

Screen Shot 2021-09-29 at 11 20 37 PM

**Comment by spydon** [comment]
> After all the tweaks, the doc now renders like this:
> 
> Screen Shot 2021-09-29 at 11 20 37 PM
> 

Very nice!!

**Comment by luanpotter** [comment]
> the purpose of this folder is not to provide build configuration

I was thinking more of a way to separate any and all "build" files from the markdowns, so someone browsing on github doesn't get lost reading some config file or css. but maybe the name isn't the best then, how about:

```
+ /doc
|--+ /sphinx
|   |-- /config
|   |-- /static
|-- actual md files
```

woudl that work?

**Review by luanpotter — APPROVED** [review]
LGTM, awesome work, thanks @st-pasha

**Comment by st-pasha** [comment]
@luanpotter  We could make nested folders like you're suggesting. The only question then is what do we do with `images/`? Conceptually, they're part of the static assets, so should go into that folder. But at the same time, they are very close to being the "content", and in fact, aren't really sphinx-specific. I guess we could have a rule that those images that are embedded into the page(s) will go into the root `images/` folder, while those images/assets that are needed specifically for sphinx would go into the `sphinx/images` folder? It is possible that at some point we'd want to add custom fonts (if Arial is too boring), or scripts (for example in order to embed dart samples into the page) -- but these won't be renderable on GitHub anyway.

Btw, sphinx uses `_` in front of a name to signify "special" folders, resembling the convention for private variables in a class. Do we want to stick to this convention (so it'd be `_sphinx/` instead `sphinx/`)?

**Review by erickzanardo — APPROVED** [review]

**Comment by luanpotter** [comment]
@st-pasha totally agree with your distinction between two kinds of images folders, one for actual content and one for sphinx related stuff!

for the `_`, if it's a normal sphinx convention, I am ok with naming it `_sphinx`!

**Review by luanpotter — APPROVED** [review]
LGTM

**Pull request merged**